### PR TITLE
ObjMure2 OK and documented

### PR DESCRIPTION
**Pull request by louist103**
Sorry it took so long. Wanted to get the docs right. Used a C debugger to be extra sure.

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 1]
```diff
@@ -30,43 +47,209 @@ ActorInit Obj_Mure2_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_809613FC.s")
+static s16 sActorIds[] = {
+    ACTOR_EN_KUSA,
+    ACTOR_EN_KUSA,
+    ACTOR_EN_ISHI,
+};
+typedef struct Mure2SpawnInfo {
+    s16 dist;
+    s16 angle;
+} Mure2SpawnInfo;
```
Comment (mzxrules) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:76]: I should point out that there is a VecPolar type I added to z64math.h in my grass documentation pr that takes a float for distance instead of a s16

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 2]
```diff
@@ -30,43 +47,209 @@ ActorInit Obj_Mure2_InitVars = {
...
+
+static Mure2SpawnInfo sScatteredBushSpawnInfo[] = {
+    { 0x0028, 0x0666 }, { 0x0028, 0x2CCC }, { 0x0028, 0x5999 }, { 0x0028, 0x8666 },
+    { 0x0014, 0xC000 }, { 0x0050, 0x1333 }, { 0x0050, 0x4000 }, { 0x0050, 0x6CCC },
+    { 0x0050, 0x9333 }, { 0x0050, 0xACCC }, { 0x0050, 0xC666 }, { 0x003C, 0xE000 },
```
Comment (mzxrules) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:81]:
```suggestion
    { 40, 0x0666 }, { 40, 0x2CCC }, { 40, 0x5999 }, { 40, 0x8666 },
    { 20, 0xC000 }, { 80, 0x1333 }, { 80, 0x4000 }, { 80, 0x6CCC },
    { 80, 0x9333 }, { 80, 0xACCC }, { 80, 0xC666 }, { 60, 0xE000 },
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.h [hunk 3]
```diff
@@ -10,7 +10,10 @@ typedef void (*ObjMure2ActionFunc)(struct ObjMure2*, PlayState*);
 typedef struct ObjMure2 {
     /* 0x000 */ Actor actor;
     /* 0x144 */ ObjMure2ActionFunc actionFunc;
-    /* 0x148 */ char unk_148[0x38];
+    /* 0x148 */ Actor* actors[12];
```
Comment (mzxrules) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.h:13]:
```suggestion
    /* 0x148 */ Actor* actors[OBJ_MURE2_CHILD_MAX];
```

Think this should be defined as a constant and used appropriately; the 12 in sChildCounts shouldn't be replaced with this constant since number of bushes should be treated as being independent of child max.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 4]
```diff
@@ -10,14 +10,35 @@
 
 #define THIS ((ObjMure2*)thisx)
 
+#define OBJ_MURE2_GET_CHILD_TYPE(thisx) ((thisx)->params & 3)
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:13]: Generally we just put all getters in the header file.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 5]
```diff
@@ -10,14 +10,35 @@
...
+
+#define OBJ_MURE2_CHILD_COUNT_BUSH_RING 9
+#define OBJ_MURE2_CHILD_COUNT_BUSH_SCATTERED 12
+#define OBJ_MURE2_CHILD_COUNT_ROCK_RING 8
+
+typedef enum Mure2ChildType {
+    OBJMURE2_CHILDTYPE_BUSH_RING,
+    OBJMURE2_CHILDTYPE_BUSH_SCATTERED,
+    OBJMURE2_CHILDTYPE_ROCK_RING
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:22]:
```suggestion
    /* 0 */ OBJMURE2_CHILDTYPE_BUSH_RING,
    /* 1 */ OBJMURE2_CHILDTYPE_BUSH_SCATTERED,
    /* 2 */ OBJMURE2_CHILDTYPE_ROCK_RING
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 6]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+static s16 sActorIds[] = {
+    ACTOR_EN_KUSA,
+    ACTOR_EN_KUSA,
+    ACTOR_EN_ISHI,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_809613FC.s")
+static VecPolarS sScatteredBushSpawnInfo[] = {
+    { 40, 0x0666 }, { 40, 0x2CCC }, { 40, 0x5999 }, { 40, 0x8666 }, { 20, 0xC000 }, { 80, 0x1333 },
+    { 80, 0x4000 }, { 80, 0x6CCC }, { 80, 0x9333 }, { 80, 0xACCC }, { 80, 0xC666 }, { 60, 0xE000 },
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:80]:
```suggestion
    { 40, 0x0666 }, { 40, 0x2CCC }, { 40, 0x5999 }, { 40, -0x799A }, { 20, -0x4000 }, { 80, 0x1333 },
    { 80, 0x4000 }, { 80, 0x6CCC }, { 80, -0x6CCD }, { 80, -0x5334 }, { 80, -0x399A }, { 60, -0x2000 },
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 7]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+    OBJ_MURE2_CHILD_COUNT_BUSH_SCATTERED, // OBJMURE2_CHILDTYPE_BUSH_SCATTERED
+    OBJ_MURE2_CHILD_COUNT_ROCK_RING,      // OBJMURE2_CHILDTYPE_ROCK_RING
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_809613E8.s")
+static s16 sActorIds[] = {
+    ACTOR_EN_KUSA,
+    ACTOR_EN_KUSA,
+    ACTOR_EN_ISHI,
+};
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:76]:
```suggestion
static s16 sActorIds[] = {
    ACTOR_EN_KUSA, // OBJMURE2_CHILDTYPE_BUSH_RING
    ACTOR_EN_KUSA, // OBJMURE2_CHILDTYPE_BUSH_SCATTERED
    ACTOR_EN_ISHI, // OBJMURE2_CHILDTYPE_ROCK_RING
};
```
Enum comments seem relevant/helpful here as well.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 8]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+        (pos + i)->z += 80.0f * Math_CosS(i * 0x2000);
+    }
+}
+
+void ObjMure2_GetChildParams(s16* childParams, ObjMure2* this) {
+    s32 childType;
+    s32 temp_a2;
+
+    childType = OBJ_MURE2_GET_CHILD_TYPE(&this->actor);
+    temp_a2 = (this->actor.params >> 8) & 0x1F;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:133]:
```suggestion
    s32 childType = OBJ_MURE2_GET_CHILD_TYPE(&this->actor);
    s32 temp_a2 = (this->actor.params >> 8) & 0x1F;
```
Can be defined where declared. Also should add a Getter for this other use of params unpacking.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 9]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+    if (childType == OBJMURE2_CHILDTYPE_ROCK_RING) {
+        *childParams = temp_a2 << 4;
+        return;
+    }
+    *childParams = temp_a2 << 8;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:138]:
```suggestion
        *childParams = temp_a2 << 4;
    } else {
        *childParams = temp_a2 << 8;
    }
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 10]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+}
+void ObjMure2_SpawnChildren(ObjMure2* this, PlayState* play) {
+    Vec3f* pos;
+    s32 childType;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:142]:
```suggestion
    s32 childType = OBJ_MURE2_GET_CHILD_TYPE(&this->actor);
```
Same here.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 11]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+            this->actors[i] = NULL;
+        }
+    }
+}
+
+void ObjMure2_Init(Actor* thisx, PlayState* play) {
+    ObjMure2* this = THIS;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    if (play->csCtx.state != 0) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:203]:
```suggestion
    if (play->csCtx.state != CS_STATE_IDLE) {
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 12]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+        this->actor.flags &= ~ACTOR_FLAG_10;
+        ObjMure2_KillChildren(this, play);
+        ObjMure2_SetupWaitForPlayerInRange(this);
+    }
+}
+
+void ObjMure2_Update(Actor* thisx, PlayState* play) {
+    ObjMure2* this = THIS;
+
+    if (play->csCtx.state == 0) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:248]:
```suggestion
    if (play->csCtx.state == CS_STATE_IDLE) {
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.h [hunk 13]
```diff
@@ -7,10 +7,15 @@ struct ObjMure2;
...
+#define OBJ_MURE2_CHILD_MAX 12
+
 typedef struct ObjMure2 {
     /* 0x000 */ Actor actor;
     /* 0x144 */ ObjMure2ActionFunc actionFunc;
-    /* 0x148 */ char unk_148[0x38];
+    /* 0x148 */ Actor* actors[OBJ_MURE2_CHILD_MAX];
+    /* 0x178*/  u16 spawnFlags;
+    /* 0x17C */ f32 rangeMultiplier;
+
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.h:18]:
```suggestion
```
Extra new line

**Review by hensldm — COMMENTED** [review]
Also looks like you might be having bss issues in the latest commit.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 14]
```diff
@@ -10,14 +10,35 @@
...
+
+#define OBJ_MURE2_CHILD_COUNT_BUSH_RING 9
+#define OBJ_MURE2_CHILD_COUNT_BUSH_SCATTERED 12
+#define OBJ_MURE2_CHILD_COUNT_ROCK_RING 8
+
+typedef enum Mure2ChildType {
+    OBJMURE2_CHILDTYPE_BUSH_RING,
+    OBJMURE2_CHILDTYPE_BUSH_SCATTERED,
+    OBJMURE2_CHILDTYPE_ROCK_RING
+} Mre2ChildType;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:23]:
```suggestion
} Mure2ChildType;
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c (hunk 5 above)
Comment (AngheloAlf) [reply src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:22]: A `_MAX` would be nice too, so it can be used as the size for all the arrays that are indexed with this enum

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 15]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+typedef void (*ObjMure2SpawnPosFunc)(Vec3f[12], ObjMure2*);
+static ObjMure2SpawnPosFunc sSpawnPosFuncs[] = {
+    ObjMure2_GetBushCircleSpawnPos, // OBJMURE2_CHILDTYPE_BUSH_RING
+    ObjMure2_GetBushScatteredPos,   // OBJMURE2_CHILDTYPE_BUSH_SCATTERED
+    ObjMure2_GetRocksSpawnPos,      // OBJMURE2_CHILDTYPE_ROCK_RING
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_80961490.s")
+// static InitChainEntry sInitChain[] = {
+static InitChainEntry sInitChain[] = {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:91]:
```suggestion
static InitChainEntry sInitChain[] = {
```

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c (hunk 8 above)
Comment (AngheloAlf) [reply src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:133]: You could use `Mure2ChildType` for the `childType` too

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 16]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
+    { 80, 0x4000 }, { 80, 0x6CCC }, { 80, 0x9333 }, { 80, 0xACCC }, { 80, 0xC666 }, { 60, 0xE000 },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_8096147C.s")
+typedef void (*ObjMure2SpawnPosFunc)(Vec3f[12], ObjMure2*);
+static ObjMure2SpawnPosFunc sSpawnPosFuncs[] = {
+    ObjMure2_GetBushCircleSpawnPos, // OBJMURE2_CHILDTYPE_BUSH_RING
+    ObjMure2_GetBushScatteredPos,   // OBJMURE2_CHILDTYPE_BUSH_SCATTERED
+    ObjMure2_GetRocksSpawnPos,      // OBJMURE2_CHILDTYPE_ROCK_RING
+};
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:88]: Could you move this down a bit? That way the forward declarations wouldn't be needed

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c [hunk 17]
```diff
@@ -30,43 +51,204 @@ ActorInit Obj_Mure2_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_809613FC.s")
+static VecPolarS sScatteredBushSpawnInfo[] = {
+    { 40, 0x0666 }, { 40, 0x2CCC }, { 40, 0x5999 }, { 40, 0x8666 }, { 20, 0xC000 }, { 80, 0x1333 },
+    { 80, 0x4000 }, { 80, 0x6CCC }, { 80, 0x9333 }, { 80, 0xACCC }, { 80, 0xC666 }, { 60, 0xE000 },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Mure2/func_8096147C.s")
+typedef void (*ObjMure2SpawnPosFunc)(Vec3f[12], ObjMure2*);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.c:83]: Correct me if I'm wrong, but this `12` seems to be the `OBJ_MURE2_CHILD_MAX` constant you made?
```suggestion
typedef void (*ObjMure2SpawnPosFunc)(Vec3f[OBJ_MURE2_CHILD_MAX], ObjMure2*);
```

File: Makefile [hunk 18]
```diff
@@ -65,8 +65,8 @@ endif
 
 #### Tools ####
 
-ifeq ($(shell type mips-linux-gnu-ld >/dev/null 2>/dev/null; echo $$?), 0)
-  MIPS_BINUTILS_PREFIX := mips-linux-gnu-
+ifeq ($(shell type mips64-ultra-elf-ld >/dev/null 2>/dev/null; echo $$?), 0)
```
Comment (hensldm) [inline Makefile:68]: Please revert these changes.

File: format.py [hunk 19]
```diff
@@ -15,7 +15,7 @@
 
 # clang-format, clang-tidy and clang-apply-replacements default version
 # This specific version is used when available, for more consistency between contributors
-CLANG_VER = 14
+CLANG_VER =""
```
Comment (hensldm) [inline format.py:18]: These 2

**Review by hensldm — APPROVED** [review]
Looks good to me, though a couple things need to be reverted.

File: include/z64math.h [hunk 20]
```diff
@@ -42,6 +42,11 @@ typedef struct {
     /* 0x8 */ s32 z;
 } Vec3i; // size = 0xC
 
+typedef struct {
+    /* 0x0 */ s16 distance;
+    /* 0x4 */ s16 angle;
+} VecPolarS; // size = 0x8
```
Comment (hensldm) [inline include/z64math.h:48]: Missed this previously. Offset and size comments are incorrect.

File: src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.h [hunk 21]
```diff
@@ -7,10 +7,16 @@ struct ObjMure2;
...
+#define OBJ_MURE2_CHILD_MAX 12
+#define OBJ_MURE2_GET_CHILD_TYPE(thisx) ((thisx)->params & 3)
+#define OBJ_MURE2_GET_CHILD_UPPER_PARAMS(thisx) ((((thisx)->params) >> 8) & 0x1F)
+
 typedef struct ObjMure2 {
     /* 0x000 */ Actor actor;
     /* 0x144 */ ObjMure2ActionFunc actionFunc;
-    /* 0x148 */ char unk_148[0x38];
+    /* 0x148 */ Actor* actors[OBJ_MURE2_CHILD_MAX];
+    /* 0x178*/  u16 spawnFlags;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Mure2/z_obj_mure2.h:18]:
```suggestion
    /* 0x178 */ u16 spawnFlags;
```

**Review by AngheloAlf — APPROVED** [review]

**Review by engineer124 — APPROVED** [review]

**Review by tom-overton — APPROVED** [review]
